shopify_integration/oauth.py

```diff
@@ -0,0 +1,170 @@
+		},
+		fields=["name", "connected_app"],
+	)
+
+	if not shops:
+		frappe.throw(_(f"No Shopify Settings found for {kwargs.get('shop')}"))
+
+	shopify_settings = shops[0]
+
+	# remove Frappe's argument to properly validate HMAC
+	kwargs.pop("cmd")
+
+	auth_url = initiate_web_application_flow(shopify_settings)
+	frappe.local.response["type"] = "redirect"
+	frappe.local.response["location"] = auth_url
+
+
+@frappe.whitelist()
+def initiate_web_application_flow(settings: Union["ShopifySettings", str]):
+	"""Return an authorization URL for the user, and save the state in Token Cache."""
+
+	if isinstance(settings, str):
+		settings: Dict = json.loads(settings)
+
+	shopify_settings: "ShopifySettings" = frappe.get_doc(
+		"Shopify Settings", settings.get("name")
+	)
+	connected_app: "ConnectedApp" = frappe.get_doc(
+		"Connected App", settings.get("connected_app")
+	)
+
+	ShopifySession.setup(
+		api_key=connected_app.client_id,
+		secret=connected_app.get_password("client_secret"),
+	)
+
+	state = binascii.b2a_hex(os.urandom(15)).decode("utf-8")
+	shopify_session = ShopifySession(
+		shopify_settings.shopify_url, shopify_settings.api_version
+	)
+	auth_url = shopify_session.create_permission_url(
+		connected_app.get_scopes(), connected_app.redirect_uri, state
+	)
+
+	# create an initial token cache for the user
+	token_cache = connected_app.get_token_cache(DEFAULT_TOKEN_USER)
+	if not token_cache:
+		token_cache = frappe.new_doc("Token Cache")
+		token_cache.user = DEFAULT_TOKEN_USER
+		token_cache.connected_app = connected_app.name
+	token_cache.state = state
+	token_cache.save(ignore_permissions=True)
+
+	return auth_url
+
+
+@frappe.whitelist(allow_guest=True)
+def callback(*args, **kwargs):
+	validate_request(*args, **kwargs)
+	connected_app, token_cache = get_oauth_details(*args, **kwargs)
+	shopify_settings = get_shopify_settings(connected_app)
+
+	if not shopify_settings:
+		frappe.throw(_(f"No Shopify Settings found for {kwargs.get('shop')}"))
+
+	# remove Frappe's argument to properly validate HMAC
+	kwargs.pop("cmd")
+
+	session = ShopifySession(shopify_settings.shopify_url, shopify_settings.api_version)
+	access_token = session.request_token(kwargs)
+	token_cache = token_cache.update_data(
+		{
+			"token_type": "Bearer",
+			"access_token": access_token,
+		}
+	)
+
+	# once the access token is fetched, register Shopify webhooks
+	shopify_settings.register_webhooks()
+
+	frappe.local.response["type"] = "redirect"
+	frappe.local.response["location"] = (
+		token_cache.get("success_uri") or connected_app.get_url()
+	)
+
+
+def validate_request(*args, **kwargs):
+	# validate request method
+	if frappe.request.method != "GET":
+		frappe.throw(_("Invalid request method: {}").format(frappe.request.method))
+
+	# validate redirect URI parameters; check if connected app is available
+	# example request path: "/api/method/path.to.method/connected_app_name"
+	path = frappe.request.path[1:].split("/")
+	if len(path) != 4 or not path[3]:
```

Comment:
Won't `path[3]` give an index error instead of returning `None` if it doesn't exist?